src/Microsoft.ML.Core/Data/ColumnType.cs

```diff
@@ -554,43 +554,47 @@ public override bool Equals(ColumnType other)
     /// class numbers, etc. For example, in multi-class classification, the label is typically
     /// a class number which is naturally a KeyType.
     ///
-    /// KeyTypes can be contiguous (the class number example), in which case they can have
-    /// a cardinality/Count. For non-contiguous KeyTypes the Count property returns zero.
-    /// Any KeyType (contiguous or not) can have a Min value. The Min value is always >= 0.
+    /// KeyTypes with a known range of values have a cardinality/Count. If the range is
+    /// unknown the Count property returns zero.
     ///
     /// Note that the representation value does not necessarily match the logical value.
-    /// For example, if a KeyType has range 1000-5000, then it has a Min of 1000, Count
-    /// of 4001, but the representational values are 1-4001. The representation value zero
-    /// is reserved to mean none/invalid.
+    /// For example, if a KeyType has range 0-5000, then it has a Count of 5001, but
+    /// the representational values are 1-5001. The representation value zero is reserved
+    /// to mean none/invalid.
     /// </summary>
     public sealed class KeyType : PrimitiveType
     {
-        private KeyType(Type type, DataKind kind, ulong min, int count, bool contiguous)
+        private KeyType(Type type, DataKind kind, ulong count)
             : base(type, kind)
         {
             Contracts.AssertValue(type);
             Contracts.Assert(kind.ToType() == type);
+            Count = count;
+        }
 
-            Contracts.CheckParam(min >= 0, nameof(min));
-            Contracts.CheckParam(count >= 0, nameof(count), "Must be non-negative.");
-            Contracts.CheckParam((ulong)count <= ulong.MaxValue - min, nameof(count));
-            Contracts.CheckParam((ulong)count <= kind.ToMaxInt(), nameof(count));
-            Contracts.CheckParam(contiguous || count == 0, nameof(count), "Must be 0 for non-contiguous");
+        public KeyType(Type type, ulong count)
+            : this(type, CheckRefRawType(type), count)
+        {
+        }
 
-            Contiguous = contiguous;
-            Min = min;
-            Count = count;
+        [BestFriend]
+        internal KeyType(Type type, int count)
+            : this(type, CheckRefRawType(type), (ulong)count)
+        {
+            Contracts.Assert(0 <= count);
         }
```

Comment:
So, just an FYI, when @eerhardt factors out IDataView, I think he *might* also be moving the canonical types. (Though I am not certain if `KeyType` is part of this. It *might* be.) If that is so, we cannot rely upon this. Indeed much of his work as we see in #2131 and #2143 and many other things already checked in deals with reversing this. So rather than introducing more internal constructors, we should I hope just use the public one? Is that unreasonable?

Reply:
Or we could have a separate public constructor since `int` is a *really* common case. But as @yaeldekel says, that should be a proper check.

---
In reply to: [248115787](https://github.com/dotnet/machinelearning/pull/2146#discussion_r248115787) [](ancestors = 248115787)

Reply:
I will keep a public constructor for `int count` with a proper check since, as you mentioned, it is really the most common case. 

Regarding the other constructor which takes `DataKind`, when possible I switched to the public constructors, but in 7 occurrences it would require a lot more work. So I decided to keep it and wait for Eric's changes to remove `DataKind`. After that it will be trivial to remove this constructor. 

---
In reply to: [248116250](https://github.com/dotnet/machinelearning/pull/2146#discussion_r248116250) [](ancestors = 248116250,248115787)

Reply:
Sounds good thank you.

---
In reply to: [248483878](https://github.com/dotnet/machinelearning/pull/2146#discussion_r248483878) [](ancestors = 248483878,248116250,248115787)